test/e2e/external_oidc_test.go

```diff
@@ -64,110 +254,172 @@ type oidcAuthResponse struct {
-		var cleanups []func()
-		kcClient, cleanups = tc.setupExternalOIDCWithKeycloak(testCtx)
-		defer test.IDPCleanupWrapper(func() {
-			for _, c := range cleanups {
-				c()
-			}
-		})()
-		t.Logf("keycloak Admin URL: %s", kcClient.AdminURL())
+	respBytes, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse well-known JWKS response body: %v", err)
 	}
 
-	group := names.SimpleNameGenerator.GenerateName("e2e-keycloak-group-")
-	err = kcClient.CreateGroup(group)
-	require.NoError(t, err)
+	var issuerJWKS jwks
+	if err := json.Unmarshal(respBytes, &issuerJWKS); err != nil {
+		return nil, fmt.Errorf("could not unmarshal JWKS: %v", err)
+	}
 
-	user := names.SimpleNameGenerator.GenerateName("e2e-keycloak-user-")
-	email := fmt.Sprintf("%s@test.dev", user)
-	password := "password"
-	err = kcClient.CreateUser(user, email, password, []string{group})
-	require.NoError(t, err)
+	return &issuerJWKS, nil
+}
 
-	httpClient := &http.Client{Transport: &http.Transport{
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-	}}
+func extractRSAPubKeyFunc(issuerJWKS *jwks) func(*jwt.Token) (any, error) {
+	return func(token *jwt.Token) (any, error) {
+		for _, key := range issuerJWKS.Keys {
+			if key.KID == token.Header["kid"] {
+				switch key.Alg {
+				case "RS256":
+					n, err := base64.RawURLEncoding.DecodeString(key.N)
+					if err != nil {
+						return nil, fmt.Errorf("could not decode N: %v", err)
+					}
+					e, err := base64.RawURLEncoding.DecodeString(key.E)
+					if err != nil {
+						return nil, fmt.Errorf("could not decode E: %v", err)
+					}
+
+					pubkey := &rsa.PublicKey{
+						N: new(big.Int).SetBytes(n),
+						E: int(new(big.Int).SetBytes(e).Int64()),
+					}
+
+					return pubkey, nil
+				}
+
+				return nil, fmt.Errorf("unexpected signing algorithm for key '%s': %s", key.KID, key.Alg)
+			}
+		}
 
-	formData := url.Values{
-		"grant_type": []string{"password"},
-		"client_id":  []string{oidcClientId},
-		"scope":      []string{"openid"},
-		"username":   []string{user},
-		"password":   []string{password},
+		return nil, fmt.Errorf("could not find an RSA key for signing use in the provided JWKS")
 	}
+}
 
-	resp, err := httpClient.PostForm(kcClient.TokenURL(), formData)
+func checkFeatureGates(t *testing.T, ctx context.Context, configClient *configclient.Clientset) {
+	featureGates, err := configClient.ConfigV1().FeatureGates().Get(ctx, "cluster", metav1.GetOptions{})
 	require.NoError(t, err)
-	defer resp.Body.Close()
-	require.Equal(t, http.StatusOK, resp.StatusCode)
 
-	data, err := io.ReadAll(resp.Body)
-	require.NoError(t, err)
-	require.NotEmpty(t, data)
+	if len(featureGates.Status.FeatureGates) != 1 {
+		// fail test if there are multiple feature gate versions (i.e. ongoing upgrade)
+		t.Fatalf("multiple feature gate versions detected")
+	} else {
+		for _, fgDisabled := range featureGates.Status.FeatureGates[0].Disabled {
+			if fgDisabled.Name == features.FeatureGateExternalOIDC {
```

Comment:
We know we are going to end up with multiple feature-gates corresponding to this functionality due to the way we need to navigate HCP having gone GA with this feature already and that impacting our ability to make API changes under a single gate.

With that in mind, maybe it is worth preemptively making this a set based check so we can easily add to the set when new feature gates are added?

The set based check may be easier to do with a flipped polarity of this check though. Checking if any of the feature gates in the set are enabled is likely easier than checking if _all_ the feature gates are in the disabled set.

Reply:
My intention was to fix this separately to avoid crashing KAS pods at all, and still have the test validate the current behavior. As discussed offline, we'll skip this test for now, and re-enable once the fix is in place.

Reply:
We discussed offline that planning for the future changes needing feature-gates is not as straightforward as I had originally thought it might be.

Feel free to ignore this suggestion.